app/api/ada/daedalusTransfer.js

```diff
@@ -52,36 +60,67 @@ export function getAddressesWithFunds(payload: {
 
 /** Generate transaction including all addresses with no change */
 export async function generateTransferTx(payload: {
-  wallet: CryptoDaedalusWallet,
-  addressesWithFunds: Array<CryptoDaedalusAddressRestored>
+  addressesWithFunds: AddressKeyMap
 }): Promise<TransferTx> {
   try {
-
-    const { wallet, addressesWithFunds } = payload;
+    const { addressesWithFunds } = payload;
 
     // fetch data to make transaction
-    const senders = addressesWithFunds.map(a => a.address);
+    const senders = Object.keys(addressesWithFunds);
     const senderUtxos = await getAllUTXOsForAddresses(senders);
     if (_.isEmpty(senderUtxos)) {
       throw new NoInputsError();
     }
-    const recoveredBalance = await getBalance(senders);
-    const inputWrappers: Array<DaedalusInputWrapper> = _getInputs(senderUtxos, addressesWithFunds);
-    const inputs = inputWrappers.map(w => w.input);
+    const inputs = utxoToTxInput(senderUtxos);
 
     // pick which address to send transfer to
-    const output = await getReceiverAddress();
+    const outputAddr = await getReceiverAddress();
+    const feeAlgorithm = RustModule.Wallet.LinearFeeAlgorithm.default();
+
+    // firts build a transaction to see what the cost would be
+    const fakeTxBuilder = new RustModule.Wallet.TransactionBuilder();
+    addTxInputs(fakeTxBuilder, inputs);
+    const inputAmount = coinToBigNumber(fakeTxBuilder.get_input_total());
+    addOutput(fakeTxBuilder, outputAddr, inputAmount.toString());
+    const fee = coinToBigNumber(fakeTxBuilder.estimate_fee(feeAlgorithm));
+
+    // now build the real transaction with the fees taken into account
+    const realTxBuilder = new RustModule.Wallet.TransactionBuilder();
+    addTxInputs(realTxBuilder, inputs);
+    const sendAmount = inputAmount.minus(fee.toString());
+    addOutput(realTxBuilder, outputAddr, sendAmount.toString());
+
+    // sanity check
+    const balance = realTxBuilder.get_balance(feeAlgorithm);
+    if (!balance.is_zero()) {
+      throw new GenerateTransferTxError();
+    }
```

Comment:
Could it be possible that reducing output size might reduce tx byte-size and therefore reduce fee and make balance positive? 🤔 I think in Rust code I have definitely seen a case when they give up few lovelaces in a case like this and just say it's ok to sacrifice them as fee.

Reply:
Yes, I think it's possible because of CBOR. I'm looking for Rust example.

Reply:
At the moment they only use sanity check: `if real_fee > selection_result.estimated_fees` - then it's error. But real fee might turn out to be `<` than estimated fee, then balance will be non-zero positive value.

https://github.com/input-output-hk/rust-cardano/blob/53d5e5fb8981523c124d2d1ebc2845bc6bdacfa8/cardano/src/wallet/scheme.rs#L156-L160

Reply:
We even opened a related issue in there - https://github.com/input-output-hk/rust-cardano/issues/349